### PR TITLE
Callback for `from`

### PR DESCRIPTION
**Pull request by CxRes**
Added a callback for `from`, which takes file as an argument. Allows the user to tailor the search string according to the filename.

You may want to rename the tests of fix the language in the documentation.

Also fixed the documentation for the `to` callback.

### CODE REVIEW

File: lib/helpers/make-replacements.js [hunk 1]
```diff
@@ -29,6 +29,10 @@ module.exports = function makeReplacements(contents, from, to, file) {
   //Make replacements
   from.forEach((item, i) => {
 
+    if (typeof item === 'function') {
```
Comment (adamreisnz) [inline lib/helpers/make-replacements.js]: Can you add a comment above this block to indicate if a function is given it's called with the file name?

File: package.json [hunk 2]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "replace-in-file",
-  "version": "3.2.0",
+  "version": "3.3.0",
```
Comment (adamreisnz) [inline package.json]: You don't need to bump the package version, I'll do that when I release a new version

**Comment by CxRes** [comment]
Changes made as requested...

File: README.md [hunk 3]
```diff
@@ -191,14 +192,26 @@ const options = {
...
+  to: 'bar',
+};
+```
+
 ### Using callbacks for `to`
-As the `to` parameter is passed straight to the native [String replace method](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/replace), you can also specify a callback. The following example uses a callback to convert matching strings to lowercase:
+As the `to` parameter is passed to the native [String replace method](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/replace), you can also specify a callback. This callback provides as an extra argument the name of the file in which the replacement is being performed. The following example uses a callback to convert matching strings to lowercase and appends the file extension to it:
 
 ```js
+var path = require('path');
```
Comment (adamreisnz) [inline README.md]: Sorry missed this, could you change `var` to `const` please, as the project is using ES6 throughout.

Comment (adamreisnz) [reply README.md]: Actually maybe remove the `path` reference altogether, as I'd like to keep the examples as simple as possible, without external dependencies.

File: README.md [hunk 4]
```diff
@@ -191,14 +192,26 @@ const options = {
...
 const options = {
   files: 'path/to/file',
-  from: /SomePattern[A-Za-z-]+/g,
-  to: (match) => match.toLowerCase(),
+  from: /SomePattern([A-Za-z-]+)/g,
+  to: (match, p1, offset, string, file)  => `${match.toLowerCase()}${path.extname(file)}`,
```
Comment (adamreisnz) [inline README.md]: Let's revert this example back to how it was to keep it simple, and instead create a second example for the `file` parameter usage.

Comment (CxRes) [reply README.md]: Ok, will fix these later in the day!

**Comment by CxRes** [comment]
On the issue of `file` argument: Could we consider making `file` the first argument in 4.0? The problem with `String.prototype.replace` is that it produces an arbitrary number of matches and hence detecting file requires some extra logic on part of the user. (I can see the other side of the argument that most users will never use file). I am open to suggestion and can make an issue if you wish...

**Comment by adamreisnz** [comment]
You could use rest parameters to get around that, e.g.

```
(match, ...rest) => {
  const file = rest.pop();
}
```

File: README.md (hunk 4 above)
Comment (CxRes) [reply README.md]: Changes made as requested!

**Comment by CxRes** [comment]
Just from an API point of view I had another idea. Instead of adding an extra argument `file`, which spoils the purity of the replace function, we can have a single callback with file argument for the `to` and `from` options which return the array of String/RegExp/Function. Something like this:

```js
{
  from: (file) => [`SomeStringThatDependsOn${file}`, ...]
  to: (file) => [(match) => `SomethingElseThatDependsOn${file}`, ...]
}
```

Just an idea :smile: !

**Pull request merged**